### PR TITLE
37318 supervisor manage subjects

### PR DESCRIPTION
**Pull request by duet-ngyen**

### CODE REVIEW

File: app/assets/stylesheets/supervisor/subjects.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the supervisor/subjects controller here.
```
Comment (ThuBM) [inline app/assets/stylesheets/supervisor/subjects.scss]: remove this file

File: app/controllers/supervisor/subjects_controller.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      flash[:notice] = t"application.notice.subject_updated"
+      redirect_to supervisor_subjects_path @subjects
+    else
+      render "edit"
+    end
+  end
+
+  def destroy
+    if @subject.destroy
+      flash[:alert] = t"application.notice.subject_deleted"
```
Comment (ThuBM) [inline app/controllers/supervisor/subjects_controller.rb]: add space after t

File: app/controllers/supervisor/subjects_controller.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    else
+      flash[:alert] = t"application.notice.subject_not_deleted"
```
Comment (ThuBM) [inline app/controllers/supervisor/subjects_controller.rb]: add space after t

File: app/controllers/supervisor/subjects_controller.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    end
+    redirect_to supervisor_subjects_path
+  end
+
+  private
+  def load_subject
+    @subject = Subject.find params[:id]
```
Comment (ThuBM) [inline app/controllers/supervisor/subjects_controller.rb]: find_by id: params[:id]

File: app/helpers/supervisor/subjects_helper.rb [hunk 5]
```diff
@@ -0,0 +1,2 @@
+module Supervisor::SubjectsHelper
```
Comment (ThuBM) [inline app/helpers/supervisor/subjects_helper.rb]: remove this file

File: app/views/supervisor/subjects/edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<%= render "form" %>
+<%= link_to t("application.button.cancel"),
+    supervisor_subject_path(@subject), class: "btn btn-default margintop" %>
```
Comment (ThuBM) [inline app/views/supervisor/subjects/edit.html.erb]: indent

File: app/views/supervisor/subjects/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+<%= paginate @subjects %>
+  <% @subjects.each do |subject| %>
+    <div id="rcorners2" class="margintop">
+      <div class="col-sm-11 bor-right">
+        <h1><%= link_to subject.title, supervisor_subject_path(subject.id) %></h1>
+        <p><%= subject.description %></p>
+      </div>
+      <div class="col-sm-1">
+        <%= link_to t("application.button.delete"), supervisor_subject_path(subject),
+            class: "btn btn-danger",
```
Comment (ThuBM) [inline app/views/supervisor/subjects/index.html.erb]: indent

File: app/views/supervisor/subjects/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+            method: :delete,
+            data: {confirm: t("application.message.delete_confirm")} %>
+      </div>
+      <div style="clear:both"></div>
```
Comment (ThuBM) [inline app/views/supervisor/subjects/index.html.erb]: shouldn't write style in html, you should write in css file and add class in here

File: app/views/supervisor/subjects/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,9 @@
+<% provide :title, @subject.title %>
+<div id="rcorners2" class="margintop">
+  <h1><%= @subject.title %></h1>
+  <p><%= @subject.description %></p>
+</div>
+<%= link_to t("application.button.edit"),
+  edit_supervisor_subject_path(@subject), class: "btn btn-primary margintop" %>
+<%= link_to t("application.button.back"),
+    supervisor_subjects_path, class: "btn btn-default margintop" %>
```
Comment (ThuBM) [inline app/views/supervisor/subjects/show.html.erb]: indent

File: app/assets/javascripts/supervisor/subjects.coffee [hunk 10]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
```
Comment (ThuBM) [inline app/assets/javascripts/supervisor/subjects.coffee]: remove this file

File: app/views/supervisor/subjects/_search_form.html.erb [hunk 11]
```diff
@@ -0,0 +1,10 @@
+<%= search_form_for @search, url: search_supervisor_subjects_path do |f| %>
+  <%= f.label t("application.title.search_by_title") %>
+  <div class="input-group">
+    <%= f.search_field :title_cont, class: "form-control",
+      placeholder:" Search for..." %>
```
Comment (ThuBM) [inline app/views/supervisor/subjects/_search_form.html.erb]: add space after placehoder:

File: app/views/supervisor/subjects/index.html.erb [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+<%= paginate @subjects %>
+<% @subjects.each do |subject| %>
+  <div id="rcorners2" class="margintop">
+    <div class="col-sm-11 bor-right">
+      <h1><%= link_to subject.title, supervisor_subject_path(subject.id) %></h1>
+      <p><%= subject.description %></p>
+    </div>
+    <div class="col-sm-1">
+      <%= link_to t("application.button.delete"), supervisor_subject_path(subject),
+          class: "btn btn-danger",
```
Comment (ThuBM) [inline app/views/supervisor/subjects/index.html.erb]: indent

File: app/controllers/supervisor/subjects_controller.rb [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+  end
+
+  def index
+    @search = Subject.ransack params[:q]
+    @subjects = Kaminari.paginate_array(@search.result).page(params[:page]).per 10
+  end
+
+  def new
+    @subject = Subject.new
+    2.times do
```
Comment (ThuBM) [inline app/controllers/supervisor/subjects_controller.rb]: 2.times {@subject.tasks.build}

File: app/views/supervisor/subjects/_form.html.erb [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+  <%= render "shared/error_messages", object: f.object %>
+  <div id="rcorners2">
+    <%= f.label :title, t("application.title.subject_title").upcase %><br>
+    <%= f.text_field :title, class: "form-control" %><br>
+    <%= f.label :description, t("application.title.subject_description").upcase %><br>
+    <%= f.text_area :description, rows: 10, class: "form-control" %>
+  </div>
+  <div class="form-group" >
+    <%= f.fields_for :tasks do |builder| %>
+      <%= render "task_form", :f => builder %>
```
Comment (ThuBM) [inline app/views/supervisor/subjects/_form.html.erb]: f: builder

File: app/views/supervisor/subjects/show.html.erb [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+</div>
+<% unless @subject.tasks.empty? %>
+  <div id="rcorners3" class="margintop task-form">
+    <h2><%= t("tasks_list") %></h2>
+    <ol>
+      <% @subject.tasks.each do |task| %>
+      <li>
+        <h4><strong><%= t("application.title.task_title") %></strong><%= task.title %></h4>
+        <p><em><strong><%= t("description") %></strong><%= task.description %></em></p>
+        <p><%= t("status") %><%= task.status %></p>
```
Comment (ThuBM) [inline app/views/supervisor/subjects/show.html.erb]: remove ()

File: app/controllers/supervisor/subjects_controller.rb [hunk 16]
```diff
@@ -0,0 +1,57 @@
+class Supervisor::SubjectsController < ApplicationController
+  before_action :load_subject, except: [:index, :new, :create]
+
+  def show
+  end
+
+  def index
+    @search = Subject.ransack params[:q]
+    @subjects = Kaminari.paginate_array(@search.result).page(params[:page]).per 10
```
Comment (dieunb) [inline app/controllers/supervisor/subjects_controller.rb]: Sử dụng cách dành cho collection thay vì array xem

File: app/models/subject.rb [hunk 17]
```diff
@@ -4,4 +4,6 @@ class Subject < ActiveRecord::Base
   has_many :course_subjects
   has_many :courses, through: :course_subjects
   has_many :tasks
+  accepts_nested_attributes_for :tasks,
+    :reject_if => lambda { |a| a[:title].blank?}, allow_destroy: true
```
Comment (dieunb) [inline app/models/subject.rb]: convention nhé `reject_if: lambda { |attribute| attribute[:title].blank?},`

**Pull request merged**